### PR TITLE
Search operation test flakiness fix

### PR DESCRIPTION
**Pull request by MaciejMierzwa**
### Description
test PR 
### Issues Resolved
- https://github.com/opensearch-project/security/issues/3426
- https://github.com/opensearch-project/security/issues/2141
- https://github.com/opensearch-project/security/issues/2169

Similar to this task: https://github.com/opensearch-project/security/issues/1917
From what I've noticed some logs are duplicated on faster machines. During test creation audit logging results were added based on actual results produced by tests. Now if there are slower moments where logging produces non-duplicated logs, those were marked as failed. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 1]
```diff
@@ -999,8 +1003,8 @@ public void shouldIndexDocumentInBulkRequest_partiallyPositive() throws IOExcept
         auditLogsRule.assertExactlyOne(userAuthenticated(LIMITED_WRITE_USER).withRestRequest(POST, "/_bulk"));
         auditLogsRule.assertExactlyOne(grantedPrivilege(LIMITED_WRITE_USER, "BulkRequest"));
         auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "CreateIndexRequest"));
-        auditLogsRule.assertExactly(6, auditPredicate(INDEX_EVENT).withEffectiveUser(LIMITED_WRITE_USER));
```
Comment (peternied) [inline src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1002]: To me its a bug if we don't have a consistent count.  If there is some kind of retry or reissuing or other dynamic contribution to the test, we should be able to measure it and verify that its behaving consistently with these assertions.

For example we might want to change `waitForSnapshotCreation` so it returns the number of calls issued, this might be causing the inconsistency, then we can use that value to indicate how many events are in the audit log for those specific scenarios.

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java [hunk 2]
```diff
@@ -776,6 +777,18 @@ private boolean getOrReplaceAllIndices(final Object request, final IndicesProvid
                 return false;
             }
             ((CreateIndexRequest) request).index(newIndices.length != 1 ? null : newIndices[0]);
+        } else if (request instanceof ResizeRequest) {
+            // clone or shrink operations
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:781]: why is this change needed? Is it only for testing?

File: src/integrationTest/java/org/opensearch/test/framework/cluster/LocalCluster.java [hunk 3]
```diff
@@ -531,4 +533,7 @@ public TestCertificates getTestCertificates() {
         return testCertificates;
     }
 
+    public void setPrimaryNode(String primaryNode) {
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/test/framework/cluster/LocalCluster.java:536]: why need primary Node?

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java (hunk 1 above)
Comment (DarshitChanpura) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1002]: +1. Is there a way we can make this more concrete?

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java (hunk 2 above)
Comment (MaciejMierzwa) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:781]: I noticed that test shrinkIndex_positive()  and others that dealt with shrinking index were ignored from the start. The request that was created in test wasn't analyzed in any way in indexResolver. After this change request is checked, authorized against index and if user has permissions, shrink operation takes place. 
I'm not sure yet if this is the case in manual testing, have to try it

**Comment by MaciejMierzwa** [comment]
Two things cause the flakiness.
1. The way requests are performed on the cluster. When HighLevelRestClient is created depending on availability, nodes are fetched in random order. Then during https calls available nodes fetched in the first step are rotated using round robin. Because of that rotation with each run different nodes might get request which causes different amount of traffic between the nodes.  
2. Another thing causing a lot of internal traffic was internal shard allocation within the cluster there might be different amount of transport messages. For some cases what seems to be helpful is to tightly control index settings and to use fewer shards. Tests containing bulk requests are example of that.

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java (hunk 1 above)
Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1002]: Seems like issue in this test was caused by transport requests between the cluster nodes. Forcing rest client to communicate through one of the managers node helps to stabilize audit logs count. (cluster.setPrimaryNode() is used for that) Other helpful thing was setting up index to use less shards.

File: src/integrationTest/java/org/opensearch/test/framework/cluster/LocalCluster.java (hunk 3 above)
Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/test/framework/cluster/LocalCluster.java:538]: During creation of high level rest client available nodes were probed, cluster managers would have highest priority, but data nodes were also checked. On top of that nodes were rotated which caused inconsistency amount of transport audit logs. This method is used to force one node to be used as rest client

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 4]
```diff
@@ -1843,28 +1943,30 @@ public void shouldRestoreSnapshot_positive() throws IOException {
...
+        auditLogsRule.assertExactly(
+            snapshotGetCount,
             userAuthenticated(LIMITED_WRITE_USER).withRestRequest(GET, "/_snapshot/test-snapshot-repository/restore-snapshot-positive")
         );
-        auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "PutRepositoryRequest"));
-        auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "CreateSnapshotRequest"));
+        auditLogsRule.assertExactly(1, grantedPrivilege(LIMITED_WRITE_USER, "PutRepositoryRequest"));
+        auditLogsRule.assertExactly(1, grantedPrivilege(LIMITED_WRITE_USER, "CreateSnapshotRequest"));
         auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "BulkRequest"));
-        auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "CreateIndexRequest"));
```
Comment (scrawfor99) [inline src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1855]: It seems like this check is dropped? Is there a reason we don't need it here? It looks like you kept it above but then stopped using it

Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1855]: Yes, it was dropped. I used IndexOperationsHelper.createIndex(cluster, WRITE_SONG_INDEX_NAME, indexSettings); to create index instead of creating it via rest API. It was causing problems with different amount of audit logs

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java (hunk 2 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:782]: okay, you can post the results of your manual testing here. I know you were planning to address this separately. Maybe file an issue and a follow-up PR?

File: src/integrationTest/java/org/opensearch/security/SnapshotSteps.java [hunk 5]
```diff
@@ -58,13 +60,21 @@ public CreateSnapshotResponse createSnapshot(String repositoryName, String snaps
...
+    public int waitForSnapshotCreation(String repositoryName, String snapshotName) {
+        AtomicInteger count = new AtomicInteger();
         GetSnapshotsRequest getSnapshotsRequest = new GetSnapshotsRequest(repositoryName, new String[] { snapshotName });
-        Awaitility.await().alias("wait for snapshot creation").ignoreExceptions().until(() -> {
-            GetSnapshotsResponse snapshotsResponse = snapshotClient.get(getSnapshotsRequest, DEFAULT);
-            SnapshotInfo snapshotInfo = snapshotsResponse.getSnapshots().get(0);
-            return SnapshotState.SUCCESS.equals(snapshotInfo.state());
-        });
+        Awaitility.await()
+            .pollDelay(250, TimeUnit.MILLISECONDS)
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/SnapshotSteps.java:67]: will these values hold or be stable?

File: src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java (hunk 2 above)
Comment (MaciejMierzwa) [reply src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java:782]: I was able to reprocude the error by manual testing, also found old task documenting the issue. Here's the task: https://github.com/opensearch-project/security/issues/2141
Here's PR where I cherry picked fix changes from this branch: https://github.com/opensearch-project/security/pull/3716
All shrink operation/resize tests are un-ignored and are passing, I'll cleanup this branch to not push same changes 2 times.

File: src/integrationTest/java/org/opensearch/security/SnapshotSteps.java (hunk 5 above)
Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/security/SnapshotSteps.java:67]: I put this part in (pollDelay()) on earlier iteration of fixing tests dealing with snapshot creation. I found that introducing a little delay before performing HTTP check call is beneficial. At this point it shouldn't really matter, I refactored the method to return a number of calls performed, so after snapshot is verified it doesn't matter if state was checked once or couple times. Asserts check returned number of calls from this method against audit messages logged.
I could remove it and check again does it passes ci correctly without these changes (it should). The default poll delay and poll interval is 100 milliseconds.
Let me know what do you think.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -110,6 +110,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
+        number: [1,2,3,4,5,6,7,8,9,10]
```
Comment (MaciejMierzwa) [inline .github/workflows/ci.yml:113]: //TODO
It needs to be reversed before a merge

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 7]
```diff
@@ -1843,28 +1869,30 @@ public void shouldRestoreSnapshot_positive() throws IOException {
                 "/_snapshot/test-snapshot-repository/restore-snapshot-positive/_restore"
             )
         );
-        auditLogsRule.assertExactlyOne(userAuthenticated(LIMITED_WRITE_USER).withRestRequest(POST, "/restored_write_song_index/_count"));
+        auditLogsRule.assertAtLeast(1, userAuthenticated(LIMITED_WRITE_USER).withRestRequest(POST, "/restored_write_song_index/_count"));
```
Comment (scrawfor99) [inline src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1872]: Can we not assert an exact for this?

**Review by scrawfor99 — COMMENTED** [review]
Overall looks good to me just one comment on a test assertion that I hope we can swap to an exact.

File: src/integrationTest/java/org/opensearch/security/SnapshotSteps.java (hunk 5 above)
Comment (DarshitChanpura) [reply src/integrationTest/java/org/opensearch/security/SnapshotSteps.java:67]: If the tests seem to be stable with this delay, I don't mind keeping it. I would like to be sure that this is not bound to change inconsistently in future as it is a magic number being introduced here.

Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/security/SnapshotSteps.java:67]: It will be consistent thanks to the number of calls returned.

File: src/main/java/org/opensearch/security/httpclient/HttpClient.java [hunk 8]
```diff
@@ -33,6 +33,7 @@
 import javax.net.ssl.SSLEngine;
```
Comment (peternied) [inline src/main/java/org/opensearch/security/httpclient/HttpClient.java:33]: Please revert these changes

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 9]
```diff
@@ -1138,14 +1147,35 @@ public void shouldDeleteDocumentInBulk_positive() throws IOException {
...
         auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "BulkRequest"));
-        auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "CreateIndexRequest"));
-        auditLogsRule.assertExactly(4, grantedPrivilege(LIMITED_WRITE_USER, "PutMappingRequest"));
-        auditLogsRule.assertExactly(6, auditPredicate(INDEX_EVENT).withEffectiveUser(LIMITED_WRITE_USER));
+        auditLogsRule.assertExactly(4, auditPredicate(null).withLayer(AuditLog.Origin.TRANSPORT));
+        auditLogsRule.assertAtLeastTransportMessages(4, grantedPrivilege(LIMITED_WRITE_USER, "PutMappingRequest"));
+        auditLogsRule.assertAtLeastTransportMessages(4, auditPredicate(INDEX_EVENT).withEffectiveUser(LIMITED_WRITE_USER));
+    }
+
+    private static void getAndSetPrimaryNode() throws IOException {
```
Comment (peternied) [inline src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1155]: We should randomly issue requests on primary and non-primary nodes.  Instead of doing this work, lets disable audit logging of transport requests - IMO they aren't important.  Then we should always have the same deterministic number of requests.

Note; if these were `audit log test` I'd have other thoughts, but that isn't the case.

**Review by peternied — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java (hunk 9 above)
Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1155]: What do you think about removing this part and starting to rotate nodes, but also leaving assertAtLeastTransportMessages() part of logs verification?

**Review by peternied — APPROVED** [review]

File: src/integrationTest/java/org/opensearch/test/framework/audit/AuditLogsRule.java [hunk 10]
```diff
@@ -122,16 +135,35 @@ private void whenTimeoutOccurs(String methodName) {
...
         TestRuleAuditLogSink.registerListener(this);
     }
 
     public void onAuditMessage(AuditMessage auditMessage) {
-        currentTestAuditMessages.add(auditMessage);
-        log.debug("New audit message received '{}', total number of audit messages '{}'.", auditMessage, currentTestAuditMessages.size());
+        if (auditMessage.getAsMap().keySet().contains("audit_transport_headers")) {
+            if (log.isDebugEnabled()) {
+                log.debug(
+                    "New transport audit message received '{}', total number of transport audit messages '{}'.",
```
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/test/framework/audit/AuditLogsRule.java:152]: Will the updated tests still pass with different cluster configurations? i.e. more data nodes added to the cluster?

I like separating out the message types here which can differentiate types of audit messages.

File: src/integrationTest/java/org/opensearch/test/framework/audit/AuditLogsRule.java [hunk 11]
```diff
@@ -122,16 +135,35 @@ private void whenTimeoutOccurs(String methodName) {
...
+                    auditMessage,
+                    currentTransportTestAuditMessages.size()
+                );
+            }
+            currentTransportTestAuditMessages.add(auditMessage);
+        } else {
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/test/framework/audit/AuditLogsRule.java:158]: nit: the only diff I see in this if-else block is the word `transport` in the log message. Can we extract and modify to pass message depending on the condition?

Comment (MaciejMierzwa) [reply src/integrationTest/java/org/opensearch/test/framework/audit/AuditLogsRule.java:158]: @DarshitChanpura there is also a different list used in each parts of if-else statement
I could refactor it into something like this:
```
public void onAuditMessage(AuditMessage auditMessage) {
    List<AuditMessage> targetList;
    String logMessage;

    if (auditMessage.getAsMap().containsKey("audit_transport_headers")) {
        targetList = currentTransportTestAuditMessages;
        logMessage = "New transport audit message received '{}', total number of transport audit messages '{}'.";
    } else {
        targetList = currentTestAuditMessages;
        logMessage = "New audit message received '{}', total number of audit messages '{}'.";
    }

    if (log.isDebugEnabled()) {
        log.debug(logMessage, auditMessage, targetList.size());
    }

    targetList.add(auditMessage);
}
``` 
It's little bit harder to read/follow what's happening so not sure about this one :/
Let me know what do you think

File: src/integrationTest/java/org/opensearch/security/SearchOperationTest.java [hunk 12]
```diff
@@ -1525,12 +1515,12 @@ public void shouldUpdateTemplate_positive() throws IOException {
...
         auditLogsRule.assertExactly(2, userAuthenticated(LIMITED_WRITE_USER).withRestRequest(PUT, "/_template/musical-index-template"));
         auditLogsRule.assertExactlyOne(userAuthenticated(LIMITED_WRITE_USER).withRestRequest(PUT, "/song-transcription-jazz/_doc/000one"));
-        auditLogsRule.assertExactly(4, grantedPrivilege(LIMITED_WRITE_USER, "PutIndexTemplateRequest"));
+        auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "PutIndexTemplateRequest"));
         auditLogsRule.assertExactlyOne(grantedPrivilege(LIMITED_WRITE_USER, "IndexRequest"));
         auditLogsRule.assertExactlyOne(grantedPrivilege(LIMITED_WRITE_USER, "BulkRequest"));
-        auditLogsRule.assertExactly(2, grantedPrivilege(LIMITED_WRITE_USER, "CreateIndexRequest"));
-        auditLogsRule.assertExactly(4, grantedPrivilege(LIMITED_WRITE_USER, "PutMappingRequest"));
-        auditLogsRule.assertExactly(10, auditPredicate(INDEX_EVENT).withEffectiveUser(LIMITED_WRITE_USER));
+        auditLogsRule.assertExactlyOne(grantedPrivilege(LIMITED_WRITE_USER, "CreateIndexRequest"));
```
Comment (scrawfor99) [inline src/integrationTest/java/org/opensearch/security/SearchOperationTest.java:1521]: Nit: How come we use `assertExactly(1...` above but `assertExactlyOne` here? Can we use only a single of these patterns?

**Review by scrawfor99 — COMMENTED** [review]
Is there a reason we cannot get a hard count for the transport messages and still have to use the "at Least" version? If so, please add it as documentation somewhere in the change otherwise please swap to the exact versions. Otherwise looks good!

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
@scrawfor99 @DarshitChanpura @cwperks Want to take another look?

**Review by scrawfor99 — APPROVED** [review]
Looks good to me! Thanks Maciej!

**Comment by peternied** [comment]
@MaciejMierzwa Could you make the backport to 2.x for this change?
